contrib/babelfishpg_tds/src/backend/tds/tdssecure.c

```diff
@@ -69,72 +89,77 @@ SslRead(BIO * h, char *buf, int size)
+ * header.  In that case, it resumes reading the header from where it left.
  */
+
 static int
 SslHandShakeRead(BIO * h, char *buf, int size)
 {
-	int			res = 0;
+	int res = 0;
 
-	if ((res = SslRead(h, buf, size)) <= 0)
-		return res;
-
-	/* very first packet of prelogin SSL handshake */
-	if (size > 0 && res > 0 && buf[0] == TDS_PRELOGIN)
+	/*
+	 * Read the TDS header if not read.  It's possible that we're reading
+	 * the header in multiple iteration.
+	 */
+	if (pkt_bytes_read < TDS_PACKET_HEADER_SIZE)
 	{
+		/* only read the bytes left in header */
+		int header_left = TDS_PACKET_HEADER_SIZE - pkt_bytes_read;
 
-		if (res < TDS_PACKET_HEADER_SIZE)
+		/* read untill we get the header at least */
+		while (header_left > 0)
 		{
-			int			remainingRead = TDS_PACKET_HEADER_SIZE - res;
-			char		tempBuf[TDS_PACKET_HEADER_SIZE];
-
-			res = 0;
+			char *cur_loc;
 
-			/*
-			 * Read the complete remaining of the header and throw away the
-			 * bytes
-			 */
-			while (res < remainingRead)
-			{
-				int			tmp_res = 0;
+			if ((res = SslRead(h, buf, header_left)) <= 0)
+				return res;
 
-				if ((tmp_res = SslRead(h, tempBuf, remainingRead - res)) <= 0)
-				{
-					return tmp_res;
-				}
-				res += tmp_res;
-			}
+			/* copy the header data read */
+			cur_loc = &(pkt_data.buf[pkt_bytes_read]);
+			memcpy(cur_loc, buf, res);
+			pkt_bytes_read += res;
 
-			/*
-			 * Read the actual data and return the res of the actual data read
-			 * Don't worry if complete read, Openssl library will take care
-			 */
-			if ((res = SslRead(h, buf, size)) <= 0)
-				return res;
+			header_left -= res;
 		}
-		else
-		{
-			int			tmp_res = 0;
-			int			i = TDS_PACKET_HEADER_SIZE;
 
-			for (i = TDS_PACKET_HEADER_SIZE; i < res; i++)
-			{
-				buf[i - TDS_PACKET_HEADER_SIZE] = buf[i];
-			}
-			res -= TDS_PACKET_HEADER_SIZE;
+		if (unlikely(pkt_data.header.pkt_type != TDS_PRELOGIN))
+			ereport(FATAL,
+					(errcode(ERRCODE_ADMIN_SHUTDOWN),
+					 errmsg("terminating connection due to unexpected ssl packet header")));
+
+		/* endian conversion is required for length */
+		pkt_data.header.length = pg_bswap16(pkt_data.header.length);
 
-			/*
-			 * Read remaining of the data. Even if the read is less than
-			 * requested size due to whatever reasons, we are good, since we
-			 * are returning the correct res value, so caller will take care
-			 * of reading the remaining data
-			 */
-			if ((tmp_res = SslRead(h, &buf[res], TDS_PACKET_HEADER_SIZE)) <= 0)
-				return tmp_res;
-			res += tmp_res;
-		}
 	}
 
+	/* At this point, we must have read the TDS header */
+	Assert(pkt_bytes_read >= TDS_PACKET_HEADER_SIZE);
+
+	/*
+	 * If SSL packet expands the current TDS prelogin packet, we need to
+	 * read some data in next iteration.  For now, just read till end of the
+	 * packet.  The caller will try to read the remaining data in the next
+	 * iteration.
+	 * XXX: We've not really seen this scenario.
+	 */
+	if (pkt_bytes_read + size > pkt_data.header.length)
```

Comment:
does this mean that data is send as multiple pre-login packets and we are reading single packet at a time? How does this align with SSL handshake data?

Reply:
As mentioned we haven't really seen this scenario. It is just an attempt to handle and log if we ever see such a scenario in SSL handshake. Previous logic was not addressing this at all.

Reply:
we are handling it with assumption that additional data must belong to next packet. Have we made sure that it is always the case for test cases we have seen so far? SSL handshake will always be done using 1 or more pre-login packets where each packet has header in the start and followed by data tracked with length field.

Reply:
As per the TDS protocol, the SSL packets will always be encapsulated in a TDS prelogin packet. It means two possible scenario:
1. One or more SSL packets are sent in a single TDS prelogin packet. We've seen this.
2. If the combined length SSL packets that should be sent together is more than TDS prelogin length, it'll be split in multiple TDS prelogin packets. We've not seen this scenario yet.

These possibilities are not specified in the protocol doc. But, we can assume it's a possibility if not explicitly mentioned in the doc. The above code tries to handle the second scenario. Another option is to throw an error as 'unhandled'. If we get an issue, we can fix the same.